### PR TITLE
Lambda extension support

### PR DESCRIPTION
**Pull request by DarcyRaynerDD**
## Description
Adds support for submitting traces via the Datadog Lambda Extension. The Datadog Lambda Extension is a special flavor of the datadog agent, the hooks into the AWS Lambda runtime. Read the draft RFC on tracing with the extension [here](https://github.com/DataDog/architecture/blob/1721621a36e8ebcecc07ea90cabb4ea4f36173a0/rfcs/serverless/lambda-extension-tracing/rfc.md) for more context.

This PR adds a sync_flush_mode to the AgentWriter, will flush immediately on write. We also change the detection mechanism for the log writer, only selecting it when we detect a serverless environment and the extension isn't present. We use a hardcoded file system path to detect the extension, (due to the way the extension API works, the extension can only be at that path).

I've added unit tests, and tested this manually as well.

## Checklist
- [ ] Added to the correct milestone.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Comment by DarcyRaynerDD** [comment]
@Kyle-Verhoog Can I get a review on this PR?

**Comment by Kyle-Verhoog** [comment]
Yes! Apologies for the delay @DarcyRaynerDD I'll have a look by EOD.

**Review by Kyle-Verhoog — COMMENTED** [review]
Looks good! I'm going to pull out the synchronous writer change to a separate PR and add a few additional test cases but I think the general approach makes sense :+1:

**Review by Yun-Kim — APPROVED** [review]

File: ddtrace/tracer.py [hunk 1]
```diff
@@ -299,12 +301,14 @@ def configure(
         self.writer.stop()
         if url:
             agent.verify_url(url)
+        sync_mode = self._is_agentless_environment()
```
Comment (Yun-Kim) [inline ddtrace/tracer.py:304]: A quick comment above explaining why being in an agentless environment triggers synchronous flush mode would be beneficial for future readers

**Review by Yun-Kim — CHANGES_REQUESTED** [review]

**Review by Yun-Kim — APPROVED** [review]

**Review by P403n1x87 — APPROVED** [review]

**Pull request merged**